### ISSUE TITLE
smf.mixedlm with small sample sizes, integer data, missing values, and skwed distribution.

### ISSUE THREAD

**Issue by TheChymera** [statsmodels/statsmodels#4801]
I'm using statsmodels' MixedLM a lot recently, and some of the data I am working with is becoming quirkier and quirkier. I am wondering whether any of the following data wold be inappropriate for the function:

1. Some of my data has as little as 2 values in some of the (e.g.) time:treatment bins (and e.g. up to 8 in others).
2. Discrete (integer) dependent variable values, e.g. event per time or feature per area counts. Technically, these could also be expressed as floats, and seem continuous, but I would still count them as discrete.
3. A lot of my data has a squeezed distribution towards 0 and a long tail extending upwards. Is this still normal enough? Is there any test I can run beforehand, or would statsmodels just warn me?

In case any of these characteristics preclude usage of MixedLM with my data, is there any other modelling tool in statsmodels (particularly one which supports repeated measures/random effects) which I can use?

**Comment by kshedden** [comment]
For #1 you might have low power but there is no reason that this would
impact a mixed linear model any more than any other type of model.

For #2, we have Poisson mixed GLM
(statsmodels.genmod.bayes_mixed_glm.PoissonMixedGLM).  It works but is
still being refined.

For #3, hard to say without more information.  There is no built-in
specification testing so you will not get a warning.  You might want to log
or sqrt transform the data and then use mixed linear model

**Comment by TheChymera** [comment]
@kshedden --- do the answers ally the same way to OLS?

**Comment by kshedden** [comment]
If you do not have repeated measures or other form of dependence in the data, you can use Poisson GLM or negative binomial GLM to handle hetroscedasticity   Transforming the response variable is another option in many cases.  It is not a panacea, but handles certain forms of heteroscedasticity well.

Other tools to look at would be GEE (for handling dependence/repeated measures), splines (for handling nonlinearities), and penalized estimation (for handling sparsity and collinearity).  There is some GAM code and GP code in branches, not sure how mature or applicable it is for you.   I mentioned the BayesMixedGLM code that handles heteroscedasticity and dependence.  It is reasonably stable now.

**Comment by TheChymera** [comment]
Sadly (or rather luckily) I do have repeated measures.

What about [PoissonBayesMixedGLM](http://www.statsmodels.org/dev/generated/statsmodels.genmod.bayes_mixed_glm.PoissonBayesMixedGLM.html)? Sounds like the combo that would address all my problems?

Also, are there any (more) usage examples for the BayesMixedGLM classes? I'd be curious to see how all of the other parameters might be used.

**Comment by kshedden** [comment]
You can use the test file to see some usage examples.  I don't have any
other examples at this point that I can share.

**Comment by TheChymera** [comment]
@kshedden as demonstrated in #4807 I am having slight difficulties with adopting a bayesian angle here. Additionally, it would be weird to analyze all my continuous data with a frequentist GLM, and the discrete one with a bayesian GLM.

So I guess what I am looking for is a frequentist Poisson Mixed GLM --- one which allowes me to run an omnibus ANOVAcontrast and post-hoc t-tests. Does statsmodels provide such a function? I have browsed the code a bit, and it seems the only poisson Mixed GLM is PoissonBayesMixedGLM --- is that correct?

**Comment by kshedden** [comment]
I think all the issues you found with the Poisson mixed GLM have been fixed
by some code I merged about a week ago.  If you find otherwise please let
me know.

Yes, this the only mixed GLM we have.

**Comment by kshedden** [comment]
A posterior standard deviation is analogous to a frequentist standard error
(i.e. it will get smaller with increasing sample size and reflects the
uncertainty about the parameter's value based on all the information in the
data).  So you would be on very solid ground using the posterior mean
divided by the posterior SD as a Z-score, and then base a t-test on it.

**Comment by TheChymera** [comment]
Ah, that does make it --- at least conceptually --- easy then. However, putting everything (especially an ANOVA) together by hand would take a while. Does statsmodels have any ANOVA or t-test bindings, which would take mean and SEM as inputs?

Also, on the conceptual level --- am I not throwing the entire robustness-to-heteroscedasticity part out the window, if I am at any step simply summarizing my data as a mean+SEM?

**Comment by kshedden** [comment]
You aren't summarizing the data as mean +/- SEM.  You are fitting the
Poisson mixed model to the data, then summarizing what you know about the
Poisson model's parameters in terms of a point estimate (posterior mean)
and uncertainty level (posterior SD).  This is similar to the frequentist
approach where you summarize what you know about model parameters using a
point estimate (MLE) and uncertainty level (standard error).

**Comment by TheChymera** [comment]
what about bindings for ANOVA and the t-test? I have tried passing the fit object from my last example to [`anova_lm`](http://www.statsmodels.org/devel/generated/statsmodels.stats.anova.anova_lm.html#statsmodels.stats.anova.anova_lm) but that fails with:

```
    anova = sm.stats.anova_lm(fit, typ=2)
  File "/usr/lib64/python3.6/site-packages/statsmodels/stats/anova.py", line 326, in anova_lm
    return anova_single(model, **kwargs)
  File "/usr/lib64/python3.6/site-packages/statsmodels/stats/anova.py", line 83, in anova_single
    robust)
  File "/usr/lib64/python3.6/site-packages/statsmodels/stats/anova.py", line 178, in anova2_lm_single
    cov = _get_covariance(model, None)
  File "/usr/lib64/python3.6/site-packages/statsmodels/stats/anova.py", line 15, in _get_covariance
    return model.cov_params()
  File "/usr/lib64/python3.6/site-packages/statsmodels/genmod/bayes_mixed_glm.py", line 858, in cov_params
    self.model.vc_names)
TypeError: can only concatenate list (not "NoneType") to list
```

**Comment by TheChymera** [comment]
For the record, it appears that's actually `scipy.stats.norm.cdf()` on `=scipy-1.1.0`. But in any case that seems to work, thank you!

How would I compute the p-value for the comparison between two factors? I assume I'd just subtract the means --- but what about the variances?

**Comment by kshedden** [comment]
Using the Laplace approximation (`fit` or `fit_map`) you should be able to create a contrast vector `d` having the same length as `result.params` containing one 1 and one -1 in positions corresponding to the two parameters you want to contrast.  Then the point estimate is `np.dot(d, result.params)`, which should give the same value as differencing the two parameter estimates.  The standard deviation for this contrast estimate is

```
np.sqrt(np.dot(d, np.dot(result.cov_params(), d)))
```
(this is a posterior standard deviation so behaves like a standard error in a frequentist analysis).

This approach will not work with the VB mean field approximation (`fit_vb`), which may in some cases give a better approximation to the posterior distribution than the Laplace approximation.  The VB method does not estimate posterior cross-covariances so cannot be used to do inference on parameter contrasts.  However you may be able to reparameterize your model so that what was once a contrast becomes a single parameter, and then the VB method can be used.

Performance trade-offs among different forms of Posterior approximation are not yet well understood.

**Comment by TheChymera** [comment]
@kshedden while `fit.summary()` shows only 5 factors (and a 6th entry for the variance term), `fit.params` gives me 39 values. Thus I get:

```
ValueError: shapes (5,) and (39,) not aligned: 5 (dim 0) != 39 (dim 0)
```

Am I doing something wrong?

**Comment by kshedden** [comment]
You get that error when doing what?

**Comment by TheChymera** [comment]
while running `np.dot`:

```
Traceback (most recent call last):
  File "sprout-countANOVA.py", line 55, in <module>
    mean = np.dot(contrast, fit.params)
ValueError: shapes (5,) and (39,) not aligned: 5 (dim 0) != 39 (dim 0)
```

Is this unexpected? Would you need a minimal working example including data?

**Comment by kshedden** [comment]
Summary only shows the fixed effects parameters and parameters for the
variance structure.  You will need to expand your contrast vector to have
length 39 (the remaining components are the realizations of the random
effects).  The fixed effects parameters come first, so if you are only
contrasting within the fixed effects, you can just pad with zeros.

**Comment by TheChymera** [comment]
@kshedden ok, I'm working on it. But you sure these z-tests are going the correct route? After all, I am dealing with discrete data (the very reason I needed the Poisson GLM), and now all my means ahave values like 0.3274 --- whereas the probability of any non-interger value is actually zero.

**Comment by TheChymera** [comment]
are you sure `fit.cov_params()` is what I need to get the standard deviations? 

`fit.cov_params()` gives me:

```
Traceback (most recent call last):
  File "sprout-countANOVA.py", line 59, in <module>
    print(fit.cov_params())
  File "/usr/lib64/python3.6/site-packages/statsmodels/genmod/bayes_mixed_glm.py", line 858, in cov_params
    self.model.vc_names)
TypeError: can only concatenate list (not "NoneType") to list
```

`fit.cov_params` gives me:

```
<bound method BayesMixedGLMResults.cov_params of <statsmodels.genmod.bayes_mixed_glm.BayesMixedGLMResults object at 0x7f433dde05c0>>
```

`fit._cov_params` is what seems to give me the values I need. Is that correct?

**Comment by TheChymera** [comment]
@kshedden I just went ahead and used `fit._cov_params`, and the p-values look about right. Overall rather smaller (so “better”) than I expected, but they vary correctly between conditions which look more similar or less similar.

At any rate, is there any way to compute an omnibus anova? Or do the holm-bonferroni corrected post-hoc t-tests suffice for multiple comparison correction. I actually think they should, but I find it customary in the literature to only reference them if an omnibus ANOVA shows the variance between groups to be significantly greater than the variance within them.

**Comment by kshedden** [comment]
The `result.cov_params` method just attaches row/column names to `result._cov_params` (if available).  So for your purposes it is fine to use `result._cov_params`.  

#4855 fixes this issue.

**Comment by kshedden** [comment]
I'm pretty confused by this:

> @kshedden ok, I'm working on it. But you sure these z-tests are going the correct route? After all, I am dealing with discrete data (the very reason I needed the Poisson GLM), and now all my means ahave values like 0.3274 --- whereas the probability of any non-interger value is actually zero.

Integer-valued distributions almost always will have non-integer expected values.  And what does this have to do with z-tests?

**Comment by TheChymera** [comment]
@kshedden I notice one more issue, namely that the factor estimates determined in this [minimal self-contained example](http://chymera.eu/debug/statsmodels/multi_longitudinal_treatment_bayes-2.zip):

```
                                        Poisson Mixed GLM Results
=========================================================================================================
                                                           Type Post. Mean Post. SD   SD  SD (LB) SD (UB)
---------------------------------------------------------------------------------------------------------
Intercept                                                     M     0.7328   0.1829                      
C(treatment, Treatment(reference='water'))[Substance1]:day    M     0.1877   0.0363                      
C(treatment, Treatment(reference='water'))[Substance2]:day    M     0.1949   0.0361                      
C(treatment, Treatment(reference='water'))[Substance3]:day    M     0.0741   0.0387                      
C(treatment, Treatment(reference='water'))[water]:day         M     0.2057   0.0360                      
subject                                                       V   -12.0000   1.0000 0.000   0.000   0.000
=========================================================================================================
Parameter types are mean structure (M) and variance structure (V)
Variance parameters are modeled as log standard deviations
```

Do not correspond to what is seen if plotting the figure:

![plot](https://user-images.githubusercontent.com/950524/43846944-caaee276-9b2f-11e8-966a-e58c11a423e5.png)


If you look at the figure, Treatments 1 and 2, as well as water, have a slope (equivalent to the interaction term with time) of around `3/3`, so `1`. The model, however, indicates interaction terms of around `0.2` --- so about one fifth. Any idea what's happening?

I have checked and the value is correctly types as `int` and the day is correctly typed as `float64`.

**Comment by TheChymera** [comment]
> This is a Bayesian analysis, so the most appropriate way to do something like this would presumably be to do some sort of Bayesian model comparison like a Bayes factor (between the model with the terms in question removed, and the model with them present).

Ok, so I understand this is no longer so straight forward. However, given that I correct the p-values for multiple comparison, is it acceptable to jump into that and comment on the significant ones without having done an omnibus ANOVA or a Bayesian model comparison?

**Comment by kshedden** [comment]
The Poisson GLM (including mixed) uses a log link.  So your results indicate that substance 1, substance 2, and water increase around 20% per day, and substance 3 increases around 7% per day.   Looks about right to me.

**Comment by kshedden** [comment]
That's probably what I would do.  But I can't guarantee that someone else
won't disagree...

**Comment by TheChymera** [comment]
@kshedden I see - is there any way to undo this log link? After all, I am looking to fit a linear  model, not an exponential function. Would this correction invalidate any of the p-value calculations discussed earlier?

**Comment by josef-pkt** [comment]
The last example is missing a `day` main effect, the slope of the reference treatment "water" is assumed to be zero.
That's way the coefficient of substance3 (smallest slope) is closes to the reference level

**Comment by TheChymera** [comment]
@josef-pkt yes, that's done by choice, I would like to estimate the absolute slope for each treatment, so as to be able to say “for **treatment** X, **value** increases by Y, daily”.

**Comment by kshedden** [comment]
There is a slope for water, it is 0.2057.  The reason that substance 3 has a small slope is because it is not increasing very fast, consistent with the graph.

**Comment by josef-pkt** [comment]
ok, sorry, my mistake reading the summary

**Comment by TheChymera** [comment]
@kshedden yes, I wasn't calling that into question at all, I was wondering why the slopes for water and the other substances were smaller than they looked on the graph --- but you clarified that. That's all ok now.

The only problem is that now I am modelling my data with an exponential function (if it is X**%** per day), and not with a liniar one. How can I get a liniar estimate? The liniar slope for water is clearly not 0.2, but more like 1.

**Comment by kshedden** [comment]
This issue started off with concerns about skewed data, that is how we got into discussing the mixed Poisson model.  Skewed, positive data are often better described through a log-scale model, especially if the magnitude of the noise increases with the mean (mean/variance relationship).

Often the mean/variance relationship and nonlinear mean structure occur together.  This is why Poisson regression is so useful (it addresses both issues).  In a non-mixed GLM, you can use a non-canonical link function to get the mean/variance model without the exponential mean structure, e.g. you could have a Poisson-like mean/variance relationship (variance proportional to the mean), while still having a linear model for the mean.  However, these non-canonical GLM's sometimes are difficult to fit, especially if the link does not satisfy a domain constraint (e.g. a linear mean model can go negative which gives undefined variances).

In Statsmodels, fixed effect GLM and GEE (an extension of GEE to dependent data) allow non-canonical links, but our mixed GLM does not.  We exploit the structure of the exponential link to facilitate fitting of the model (which is quite difficult to fit).  So you will not be able to experiment with non-canonical links in the framework you are using now.

**Comment by kshedden** [comment]
In the Poisson model, you can use log(day) instead of day as your covariate.  Then the mean part of the model is k*day^b1, where k = exp(b0), with [b0, b1] being the fixed effects intercept and fixed slope coefficient for log(day).  This would give you linear growth curves (if b1=1), or quadratic (if b1=2), or anything in between.

The model selection question is partly about the mean structure, but also partly about the distribution of the data given covariates and random effects.  If you have technical replicates (so covariates and random effects are the same) you might be able to look at sample mean versus sample variance to see if their relationship is constant or increasing.   If it is constant, use MixedLM, if it is increasing linearly, use Poisson, if it is increasing faster than linearly, you would want to use Negative Binomial or perhaps inverse Gaussian, but we don't have those as mixed (you would have to use GEE if you want to stay in statsmodels).

**Comment by TheChymera** [comment]
how would I use `log(day)` instad of `day`?

The following formuls is not properly recognized:

```
formula="value ~ Treatment : log(Q('Day Post Procedure'))"
model = PoissonBayesMixedGLM.from_formula(formula, {'subject': '0 + C(subject)'}, df)
```

**Comment by kshedden** [comment]
Something like this should work:

```
y ~ I(np.log(x))
```
